### PR TITLE
Search in main menu #1244

### PR DESCRIPTION
**Pull request by yuriy-khrebtov**
See https://github.com/jmix-framework/jmix/issues/1244

### CODE REVIEW

File: jmix-flowui/flowui-kit/src/main/java/io/jmix/flowui/kit/meta/component/StudioComponents.java [hunk 1]
```diff
@@ -2646,4 +2646,59 @@ public interface StudioComponents {
...
+            category = "Components",
+            xmlElement = "menuSearchField",
+            properties = {
+                    @StudioProperty(xmlAttribute = "alignSelf", type = StudioPropertyType.ENUMERATION,
+                            classFqn = "com.vaadin.flow.component.orderedlayout.FlexComponent$Alignment",
+                            defaultValue = "AUTO",
+                            options = {"START", "END", "CENTER", "STRETCH", "BASELINE", "AUTO"}),
+                    @StudioProperty(xmlAttribute = "ariaLabel", type = StudioPropertyType.LOCALIZED_STRING),
+                    @StudioProperty(xmlAttribute = "ariaLabelledBy", type = StudioPropertyType.LOCALIZED_STRING),
+                    @StudioProperty(xmlAttribute = "autocapitalize", type = StudioPropertyType.ENUMERATION,
```
Comment (glebfox) [inline jmix-flowui/flowui-kit/src/main/java/io/jmix/flowui/kit/meta/component/StudioComponents.java:2662]: Does it make sense having such attribute for search field?

File: jmix-flowui/flowui/src/main/java/io/jmix/flowui/component/menusearchfield/MenuSearchFieldVariant.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+package io.jmix.flowui.component.menusearchfield;
+
+import com.vaadin.flow.component.shared.ThemeVariant;
+
+/**
+ * Set of theme variants applicable for the {@link MenuSearchField} component.
+ */
+public enum MenuSearchFieldVariant implements ThemeVariant {
+
+    LUMO_SMALL("small"),
```
Comment (glebfox) [inline jmix-flowui/flowui/src/main/java/io/jmix/flowui/component/menusearchfield/MenuSearchFieldVariant.java:26]: Let's remove `LUMO_` and `MATERIAL_` prefixes. We don't follow this bizarre Vaadin naming convention in your theme variants.

File: jmix-flowui/flowui/src/main/java/io/jmix/flowui/component/menusearchfield/MenuSearchField.java [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+                    parentMenuItem.setOpened(true);
+                }
+            } else {
+                match = selfMatch;
+            }
+            return match;
+        }
+
+        protected boolean testItemMatch(MenuItem item) {
+            return StringUtils.contains(item.getTitle(), searchString);
```
Comment (glebfox) [inline jmix-flowui/flowui/src/main/java/io/jmix/flowui/component/menusearchfield/MenuSearchField.java:230]: 1. Why not `containsIgnoreCase` by defaut? 
2. Some sort of `CASE_INSENSITIVE`/`CASE_SENSITIVE` setting may be helpful.

File: jmix-flowui/flowui/src/main/java/io/jmix/flowui/component/menusearchfield/MenuSearchField.java [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+        getContent().setValue(value);
+    }
+
+    @Override
+    public String getValue() {
+        return getContent().getValue();
+    }
+
+    @Override
+    public Registration addValueChangeListener(
```
Comment (glebfox) [inline jmix-flowui/flowui/src/main/java/io/jmix/flowui/component/menusearchfield/MenuSearchField.java:145]: I have doubt that we need `ValueChangeListener`, at least the one that works simultaneously with the default one. For example, for actions when `ActionPerformedEvent` listener is added don't execute the default logic.

**Comment by glebfox** [comment]
Also, url query parameters support is needed

**Comment by yuriy-khrebtov** [comment]
> Also, url query parameters support is needed

For the moment query parameters of MainView (where MenuSearchField is supposed to be placed) are not retained during navigation so we will not provide this kind of support. Relevant issue: https://github.com/jmix-framework/jmix/issues/2531

File: jmix-flowui/flowui/src/main/java/io/jmix/flowui/component/menusearchfield/MenuSearchField.java (hunk 3 above)
Comment (yuriy-khrebtov) [reply jmix-flowui/flowui/src/main/java/io/jmix/flowui/component/menusearchfield/MenuSearchField.java]: Added search mode for case insensitive search

**Review by glebfox — CHANGES_REQUESTED** [review]

**Review by glebfox — APPROVED** [review]

**Pull request merged**